### PR TITLE
[CPU] Added faster exp routine for lower precision data types.

### PR DESCRIPTION
**Pull request by almayne**
## Purpose

Add a faster exp routine for lower precision dtypes on NEON hardware. Observed a 3-4% speedup on neoverse V1.

## Test Plan

Run
`LD_PRELOAD="/usr/lib/aarch64-linux-gnu/libtcmalloc_minimal.so.4:/usr/lib/aarch64-linux-gnu/libgomp.so.1" taskset -c 20-27 python benchmark_attention_minimal.py`

Where benchmark_attention_minimal.py copies https://github.com/neuralmagic/vllm/blob/main/benchmarks/kernels/cpu/benchmark_cpu_attn.py with:
```
    batch_size = 64
    isa = "neon"
    seq_len = 512
    iters = 50
    dtype = torch.bfloat16
    seq_lens = [(seq_len, seq_len)] * batch_size
    num_heads = (32, 8)
    head_size = 128
```

## Test Result
baseline mean (ms) / faster exp = 1.04

### CODE REVIEW

File: csrc/cpu/cpu_attn_impl.hpp [hunk 1]
```diff
@@ -985,14 +985,15 @@ class AttentionMainLoop {
 
       // process logits
       {
+        bool is_neon_f16 = (std::is_same_v<query_t, c10::BFloat16> || std::is_same_v<query_t, c10::Half>) && ISAType == ISA::NEON;
```
Comment (fadara01) [inline csrc/cpu/cpu_attn_impl.hpp]: let's keep this file backend agnostic. We should avoid introducing `is_neon_f16`  or in general changing signatures of any functions in this file.

we should also avoid runtime branches for `fast_exp` vs `fast_exp_f16`

why don't we just template `fast_exp` on the query type and ditch the `fast_exp_fp16`
We can introduce a template parameter called `is_reduced_precision` or sth which is `true` for FP16/BF16.
and then `fast_exp<false>` would use the current impl, while fast_exp<true> would use the new, fast / less-accurate impl

**Review by fadara01 — CHANGES_REQUESTED** [review]
Thanks for your work!

I think we can integrate this better with the code base, see comment.

File: csrc/cpu/cpu_arch_macros.h [hunk 2]
```diff
@@ -108,6 +108,52 @@
...
+    // Accurate within 1 ULP for BF16 for inputs in [-87.683, 88.376] & clamps
+    // inputs outside this range to 0 / inf. Implementation is similar to
+    // exp_u20, but:
+    // - uses a third degree polynomial approximation for exp(r) instead of a
+    // fifth degree one, with coefficients re-tuned.
+    // - does not split natural log (ln) into high / low parts
+    // - clamps exp(x) to 0 for x < -87.683113f and inf for x > 88.3762589f
+    // exp(x) = 2^n (1 + exp(r))
+    // r = x - n*ln2, with n = round(x/ln2)
+    // exp(r) ~ poly(r) = r + r^2 * (c3 + c2 * r)
```
[the remark this answers is not in the document]
Comment (almayne) [reply csrc/cpu/cpu_arch_macros.h]: Done

File: csrc/cpu/cpu_arch_macros.h [hunk 3]
```diff
@@ -98,13 +112,42 @@
...
+    float32x4_t q = vfmaq_n_f32(f_c3, r, f_c2);                                \
+    float32x4_t s = vaddq_f32(vdupq_n_f32(1.0f), r);                           \
+    float32x4_t p = vfmaq_f32(s, q, r2);                                       \
+    float32x4_t y =                                                            \
+        vreinterpretq_f32_u32(vaddq_u32(vreinterpretq_u32_f32(p), e));         \
+    y = vbslq_f32(lt_lower, vdupq_n_f32(0.0f), y);                             \
+    y = vbslq_f32(gt_upper, vdupq_n_f32(INFINITY), y);                         \
+    return y;                                                                  \
+    };                                                                         \
+    auto fast_exp = [&](const vec_op::FP32Vec16& vec, bool lowp)               \
```
Comment (fadara01) [inline csrc/cpu/cpu_arch_macros.h]: can we avoid having lowp as a runtime branch?
we should know at compile time whether we want to do lowp exp or full exp

File: csrc/cpu/cpu_arch_macros.h [hunk 4]
```diff
@@ -98,13 +112,42 @@
...
                         __attribute__((always_inline)) {                       \
                           float32x4x4_t result;                                \
-                          result.val[0] = neon_expf(vec.reg.val[0]);           \
-                          result.val[1] = neon_expf(vec.reg.val[1]);           \
-                          result.val[2] = neon_expf(vec.reg.val[2]);           \
-                          result.val[3] = neon_expf(vec.reg.val[3]);           \
+                          if (lowp) {                                          \
```
Comment (fadara01) [inline csrc/cpu/cpu_arch_macros.h]: can `lowp` become a template argument, and then we do: `if constexp(lowp) ...`

File: csrc/cpu/cpu_attn_impl.hpp (hunk 1 above)
Comment (almayne) [reply csrc/cpu/cpu_attn_impl.hpp]: Have resolved without templates, as discussed offline

**Review by fadara01 — COMMENTED** [review]
LGTM, I've never tested the accuracy of this beyond attention, so I'd rather we keep the MoE as is until we test the accuracy there properly

**Review by fadara01 — APPROVED** [review]
LGTM, thank you!

@bigPYJ1151 - could you take a look plz?

File: csrc/cpu/cpu_attn_impl.hpp [hunk 5]
```diff
@@ -1258,7 +1268,11 @@ class AttentionMainLoop {
                        int32_t kv_tile_token_num, float softcap_scale) {
 #ifdef DEFINE_FAST_EXP
       DEFINE_FAST_EXP
+      bool constexpr IsReducedPrecision =
+          std::is_same_v<query_t, c10::BFloat16> ||
+          std::is_same_v<query_t, c10::Half>;
```
Comment (bigPYJ1151) [inline csrc/cpu/cpu_attn_impl.hpp:1273]: Perhaps add a new flag in the ```attention_impl_t``` is a better way.

File: csrc/cpu/cpu_arch_macros.h [hunk 6]
```diff
@@ -55,14 +55,32 @@
         vec_res = _mm512_mul_ps(vec_res, vec_two);                             \
         vec_op::FP32Vec16 res(vec_res);                                        \
         return res;                                                            \
-      };
+      };                                                                       \
+      auto fast_exp_16 =                                                       \
+          [&](const vec_op::FP32Vec16& vec)                                    \
+              __attribute__((always_inline)) { return fast_exp(vec); };
```
Comment (bigPYJ1151) [inline csrc/cpu/cpu_arch_macros.h:61]: Then this fake impl is not required.

Comment (fadara01) [reply csrc/cpu/cpu_arch_macros.h:61]: Hmmm, you can get rid of this at the cost of adding ugly ISA specific ifdefs to the generic attention implementation (which is really supposed to be ISA agnostic). 

I think this is better than adding ISA specific ifdefs to attention impl.

what do you think?

Comment (bigPYJ1151) [reply csrc/cpu/cpu_arch_macros.h:61]: After checking again I think the dispatch should be moved to ARM's ```DEFINE_FAST_EXP``` macro, because it the optimization is only for ARM. The dispatch can be done in ```fast_exp``` to select use ```neon_expf``` or ```neon_expf_f16```.  In the macro defination ```query_t``` and ```q_buffer_t``` can be referenced.

Comment (almayne) [reply csrc/cpu/cpu_arch_macros.h]: fast_exp is also used in cpu_fused_moe.cpp. Neither query_t or q_buffer_t are used there. I think it would be a bit fragile to have this dependency on query_t being in scope.

Comment (bigPYJ1151) [reply csrc/cpu/cpu_arch_macros.h]: Oh... I see. Looks like there is no better way. I would prefer to add ARM platfrom macro to only apply the optimization on ARM in ```cpu_attn_impl.hpp```, and give some comments for clarity.

**Review by bigPYJ1151 — APPROVED** [review]

File: csrc/cpu/cpu_arch_macros.h (hunk 6 above)
Comment (almayne) [reply csrc/cpu/cpu_arch_macros.h]: @bigPYJ1151 you can see from the build failures that have been introduced by the last change that we have a couple of options:

1. Revert the last commit and keep the fake impl required for x86
2. Add something like the below, which is starting to look messy
```
#ifdef __aarch64__
            if constexpr (IsReducedPrecision) {
              vec = fast_exp_f16(vec);
            } else 
#endif
```

If there is a third option please let me know. I prefer option 1 but my understanding is that you would prefer option 2. If you could confirm I'll go ahead with the agreed change.

Comment (bigPYJ1151) [reply csrc/cpu/cpu_arch_macros.h]: Let's use option 2. It's more clear, even a bit messy.

Comment (fadara01) [reply csrc/cpu/cpu_arch_macros.h]: I think the original version was much better than option 2, because it doesn't introduce ISA speciific ifdefs in an isa agnostic impl and it allows you to add a faster x86 exp impl in the future.

Comment (bigPYJ1151) [reply csrc/cpu/cpu_arch_macros.h]: Haha, both are ok if there is no better method. The explict platform check may help to remind.
